pkg/gadgets/trace/exec/tracer/bpf/execsnoop.bpf.c

```diff
@@ -181,10 +223,35 @@ int ig_execve_x(struct trace_event_raw_sys_exit *ctx)
 		return 0;
 	id = bpf_get_current_pid_tgid();
 	pid = (pid_t)id;
-	event = bpf_map_lookup_elem(&execs, &pid);
+	tgid = id >> 32;
+	ret = ctx->ret;
+
+	pid_map = bpf_map_lookup_elem(&pid_by_tgid, &tgid);
+	if (!pid_map)
+		return 0;
+
+	if (ret == 0) {
+		// sys_enter_execve and sys_exit_execve might be called from different
+		// threads. We need to lookup the pid from the tgid.
+		execs_lookup_key = *pid_map;
+		event = bpf_map_lookup_elem(&execs, &execs_lookup_key);
+
+		// Remove the tgid->pid mapping. This was the last thread running a
+		// execve in parallel since there could be only one successful execve
+		// in the thread group.
+		bpf_map_delete_elem(&pid_by_tgid, &tgid);
+	} else {
+		execs_lookup_key = pid;
+		event = bpf_map_lookup_elem(&execs, &execs_lookup_key);
+		// Remove the tgid->pid mapping.
+		__atomic_add_fetch(pid_map, -pid, __ATOMIC_RELAXED);
+		if (*pid_map == 0) {
+			bpf_map_delete_elem(&pid_by_tgid, &tgid);
+		}
+	}
+
```

Comment:
I'm wondering if something like the following is more readable:

```suggestion
	// sys_enter_execve and sys_exit_execve might be called from different
	// threads. We need to lookup the pid from the tgid.
	execs_lookup_key = (ret == 0) ? *pid_map : pid;

	// Remove the tgid->pid mapping if the value reaches 0 or the execve() call was successful
	__atomic_add_fetch(pid_map, -pid, __ATOMIC_RELAXED);
	if (*pid_map == 0|| ret == 0) {
		bpf_map_delete_elem(&pid_by_tgid, &tgid);
	}

	event = bpf_map_lookup_elem(&execs, &execs_lookup_key);
```